Build failure: gforth

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

- Enter a shell with the failing package: `nix-shell -p gforth`
- `gforth-boot` fails to build due to "pointer type mismatch in conditional expression"

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/316557553

### Relevant log output

```console
error: Cannot build '/nix/store/x4fm2shfgphrq276k7wz7mxrxcwkp1an-gforth-boot-0.7.3.drv'.
       Reason: builder failed with exit code 2.
       Output paths:
         /nix/store/9lhxx8zlw29k1zs7p768an1a0lix7w38-gforth-boot-0.7.3
       Last 25 log lines:
       > signals.c:432:47: error: pointer type mismatch in conditional expression [-Wincompatible-pointer-types]
       >   432 |                ignore_async_signals ? SIG_IGN : throw_handler);
       >       |                                               ^
       > In file included from /nix/store/dj43clc5ff7jjnfmhbaj6q4q0h8kpfpm-glibc-2.40-66-dev/include/signal.h:30,
       >                  from signals.c:34:
       > signals.c:432:39: note: first expression has type 'void (*)(int)'
       >   432 |                ignore_async_signals ? SIG_IGN : throw_handler);
       >       |                                       ^~~~~~~
       > signals.c:432:49: note: second expression has type 'void (*)(void)'
       >   432 |                ignore_async_signals ? SIG_IGN : throw_handler);
       >       |                                                 ^~~~~~~~~~~~~
       > make[3]: *** [Makefile:128: signals.o] Error 1
       > make[3]: Leaving directory '/build/gforth-0.7.3/engine'
       > make[3]: Entering directory '/build/gforth-0.7.3'
       > Makefile:865: warning: ignoring prerequisites on suffix rule definition
       > Makefile:865: warning: ignoring prerequisites on suffix rule definition
       > ./preforth -p ".:~+:." -e 's" mach16b.fs"' ./kernel/main.fs -e "save-cross kernl16b.fi- /nix/store/9lhxx8zlw29k1zs7p768an1a0lix7w38-gforth-boot-0.7.3/bin/gforth-0.7.3 bye"
       > You need to configure with a gforth in $PATH to build this part
       > make[3]: *** [Makefile:716: kernl16b.fi-] Error 1
       > make[3]: Leaving directory '/build/gforth-0.7.3'
       > make[2]: *** [Makefile:812: gforth-noll] Error 2
       > make[2]: Leaving directory '/build/gforth-0.7.3'
       > make[1]: *** [Makefile:823: optgforth] Error 2
       > make[1]: Leaving directory '/build/gforth-0.7.3'
       > make: *** [Makefile:812: gforth] Error 2
       For full logs, run:
         nix log /nix/store/x4fm2shfgphrq276k7wz7mxrxcwkp1an-gforth-boot-0.7.3.drv
error: Cannot build '/nix/store/3x2fnp7r9y201a33j0pk8ahnh7sc6mwv-gforth-0.7.9_20251001.drv'.
       Reason: 1 dependency failed.
       Output paths:
         /nix/store/qn6a0p6pnqc55dn1zsfvlbki8m5bhqjl-gforth-0.7.9_20251001
error: Cannot build '/nix/store/hccpiksn29mcd4q736kzhn4k48977afy-gbforth-unstable-2025-10-08.drv'.
       Reason: 1 dependency failed.
       Output paths:
         /nix/store/mkkav9ccx8lv4hrrxmrnsdv42ndvid1v-gbforth-unstable-2025-10-08
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.63, NixOS, 26.05 (Yarara), 26.05pre919991.cad22e7d996a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2+1`
 - nixpkgs: `/nix/store/bjgc3nmjz40w51x8gl0xvccspdzlr7ns-source`

### Notify maintainers

This package has no maintainers

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/gf/gforth/boot-forth.nix
+++ b/pkgs/by-name/gf/gforth/boot-forth.nix
@@ -16,6 +16,11 @@ stdenv.mkDerivation {
     sha256 = "1c1bahc9ypmca8rv2dijiqbangm1d9av286904yw48ph7ciz4qig";
   };
 
+  patches = [
+    # Newer GCC rejects casting between function pointers omitting parameters
+    ./function-pointers.patch
+  ];
+
   buildInputs = [ m4 ];
 
   configureFlags = lib.optionals stdenv.hostPlatform.isDarwin [ "--build=x86_64-apple-darwin" ];

--- a/pkgs/by-name/gf/gforth/function-pointers.patch
+++ b/pkgs/by-name/gf/gforth/function-pointers.patch
@@ -1,0 +1,26 @@
+diff --git a/engine/signals.c b/engine/signals.c
+index b522f19..7f89bcd 100644
+--- a/engine/signals.c
++++ b/engine/signals.c
+@@ -399,7 +399,7 @@ void install_signal_handlers(void)
+ #endif
+   };
+   int i;
+-  void (*throw_handler)() = die_on_signal ? graceful_exit : signal_throw;
++  void (*throw_handler)(int) = die_on_signal ? graceful_exit : signal_throw;
+ #ifdef SIGSTKSZ 
+   stack_t sigstack;
+   int sas_retval=-1;
+diff --git a/engine/support.c b/engine/support.c
+index 021fa22..069c32f 100644
+--- a/engine/support.c
++++ b/engine/support.c
+@@ -79,7 +79,7 @@ char *tilde_cstr(Char *from, UCell size, int clear)
+ {
+   char *s1,*s2;
+   int s1_len, s2_len;
+-  struct passwd *getpwnam (), *user_entry;
++  struct passwd *getpwnam (const char *), *user_entry;
+ 
+   if (size<1 || from[0]!='~')
+     return cstr(from, size, clear);

--- a/pkgs/by-name/gf/gforth/package.nix
+++ b/pkgs/by-name/gf/gforth/package.nix
@@ -17,13 +17,13 @@ let
 in
 stdenv.mkDerivation (finalAttrs: {
   pname = "gforth";
-  version = "0.7.9_20251001";
+  version = "0.7.9_20251203";
 
   src = fetchFromGitHub {
     owner = "forthy42";
     repo = "gforth";
     rev = finalAttrs.version;
-    hash = "sha256-u9snXcFa/YYvITgMBY8FRYyyLFhHCP6hWA5ljwdKGLk=";
+    hash = "sha256-d1LU6FGnGOzL5Tdl0VLkYzHH+pUevGNCNwIqudS8emg=";
   };
 
   patches = [ ./use-nproc-instead-of-fhs.patch ];

